docs/2024/puzzles/day15.md

```diff
@@ -2,10 +2,402 @@ import Solver from "../../../../../website/src/components/Solver.js"
+object Day15:
+  def parseMoves(input: Iterator[String]): Seq[Direction] =
+    // extra flatMap because the moves may be split across multiple lines
+    input.flatMap(_.map(Direction.ofChar)).toSeq
+
+  def parse(inputFile: String): (Warehouse, Seq[Direction]) =
+    val file = io.Source.fromFile(inputFile)
+    try
+      val input = file.getLines()
+      val cells = ArraySeq.from(input
+        // an empty line separates the grid from the moves
+        .takeWhile(!_.isEmpty)
+        .map(l => ArraySeq.from(l.toCharArray()).map(WarehouseCell.ofChar)))
+      val robotRow = cells.indexWhere(_.contains(WarehouseCell.Robot))
+      val robotCol = cells(robotRow).indexOf(WarehouseCell.Robot)
+      (Warehouse(cells, robotRow, robotCol), parseMoves(input))
+    finally
+      file.close()
+```
+
+(We could save a couple traversals by detecting the robot while reading the input, using either a `var` to store its position or writing it as a fold, but the code is more elegant this way and performance is not critical.)
+
+Now, we compute the result of a single move instruction. In `getStackLengthFromTowards`, we either find the length of the stack of adjacent boxes that can successfully be pushed in the given direction (i.e., the cell after those boxes is empty) by checking just the next one and then recursing, or producing `None` if they will push against a wall or go out of bounds. The `move` method uses `getStackLengthFromTowards`, starting from the robot’s position, to update the grid.
+
+```scala
+class Warehouse:
+  /* ... */
+
+  def getStackLengthFromTowards(posRow: Int, posCol: Int, d: Direction): Option[Int] =
+    val (nextRow, nextCol) = d(posRow, posCol)
+    cells.lift(nextRow).flatMap(_.lift(nextCol))
+      .flatMap(_ match
+        case WarehouseCell.Empty => Some(0)
+        case WarehouseCell.Wall => None
+        case WarehouseCell.Box =>
+          getStackLengthFromTowards(nextRow, nextCol, d).map(_ + 1)
+        case WarehouseCell.Robot =>
+          throw AssertionError("should not reach robot in traversal"))
+
+  def move(d: Direction): Warehouse =
+    getStackLengthFromTowards(robotRow, robotCol, d) match
+      case None => this
+      case Some(stackLength) =>
+        val (newRobotRow, newRobotCol) = d(robotRow, robotCol)
+        val newCells = cells.updated(Seq(
+          ((robotRow, robotCol), WarehouseCell.Empty),
+          ((newRobotRow, newRobotCol), WarehouseCell.Robot)) ++
+          (if stackLength > 0
+           // stackLength + 1 because we want the cell after the stack
+           then Seq((d(robotRow, robotCol, stackLength + 1), WarehouseCell.Box))
+           else Seq(/* only the robot moves */)))
+        Warehouse(newCells, newRobotRow, newRobotCol)
+```
+
+Some nifty Scala features at play:
+* `.lift` lets us turn a potentially exception-throwing method (`ArraySeq.apply` out of bounds) into one that returns an `Option`, which we can further `flatMap` with the `Option`-valued wall/box/empty logic.
```

Comment:
I'm not sure this is the perfect wording, as `.lift` doesn't work on any exception-throwing method, it only (iirc) works on `PartialFunction`